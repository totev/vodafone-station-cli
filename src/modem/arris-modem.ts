import {Log} from '../logger'
import {DocsisStatus, HumanizedDocsis31ChannelStatus, HumanizedDocsisChannelStatus, Modem, DocsisChannelType, ExposedHostSettings, HostExposureSettings, Protocol} from './modem'
import {decrypt, deriveKey, encrypt} from './tools/crypto'
import {CryptoVars, extractCredentialString, extractCryptoVars, extractDocsisStatus} from './tools/html-parser'

export interface ArrisDocsisStatus {
  downstream: ArrisDocsisChannelStatus[];
  upstream: ArrisDocsisChannelStatus[];
  downstreamChannels: number;
  upstreamChannels: number;
  ofdmChannels: number;
  time: string;
}

export interface ArrisDocsisChannelStatus {
  ChannelID: string;
  ChannelType: DocsisChannelType;
  Frequency: string|number;
  Modulation: string;
  PowerLevel: string;
  SNRLevel?: string | number;
  LockStatus: string;
}

export interface SetPasswordRequest {
  AuthData: string;
  EncryptData: string;
  Name: string;
}

export interface SetPasswordResponse {
  p_status: string;
  encryptData: string;
  p_waitTime?: number;
}

interface ArrisGetHostExposureSettings {
  hostExposure: ArrisGetExposedHostSettings[];
  dhcpclient: any;
}

interface ArrisGetExposedHostSettings {
  ServiceName: string;
  MAC: string;
  Protocol: Protocol;
  StartPort: number;
  EndPort: number;
  Status: string;
  Index: string;
}

interface ArrisSetHostExposureSettings {
  hEditRule: ArrisSetExposedHostSettings[];
}

interface ArrisSetExposedHostSettings {
  name: string;
  macAddress: string;
  protocol: Protocol;
  startPort: number;
  endPort: number;
  enable: string;
  index: string;
}

export function normalizeChannelStatus(channelStatus: ArrisDocsisChannelStatus): HumanizedDocsisChannelStatus | HumanizedDocsis31ChannelStatus {
  const frequency: Record<string, number> = {}
  if (channelStatus.ChannelType === 'SC-QAM') {
    frequency.frequency = channelStatus.Frequency as number
  }

  if (['OFDM', 'OFDMA'].includes(channelStatus.ChannelType)) {
    const ofdmaFrequency = String(channelStatus.Frequency).split('~')
    frequency.frequencyStart = Number(ofdmaFrequency[0])
    frequency.frequencyEnd = Number(ofdmaFrequency[1])
  }

  const powerLevel = parseFloat(channelStatus.PowerLevel.split("/")[0]);
  const snr = parseInt(`${channelStatus.SNRLevel ?? 0}`, 10);
  return {
    channelId: channelStatus.ChannelID,
    channelType: channelStatus.ChannelType,
    modulation: channelStatus.Modulation,
    powerLevel: isNaN(powerLevel) ? -100 : powerLevel,
    lockStatus: channelStatus.LockStatus,
    snr: isNaN(snr) ? 0 : snr,
    ...frequency,
  } as HumanizedDocsisChannelStatus | HumanizedDocsis31ChannelStatus;
}

export function normalizeDocsisStatus(arrisDocsisStatus: ArrisDocsisStatus): DocsisStatus {
  const result: DocsisStatus = {
    downstream: [],
    downstreamOfdm: [],
    upstream: [],
    upstreamOfdma: [],
    time: arrisDocsisStatus.time
  }
  result.downstream = arrisDocsisStatus.downstream
    .filter(downstream => downstream.ChannelType === 'SC-QAM')
    .map(normalizeChannelStatus) as HumanizedDocsisChannelStatus[]

  result.downstreamOfdm = arrisDocsisStatus.downstream
    .filter(downstream => downstream.ChannelType === 'OFDM')
    .map(normalizeChannelStatus) as HumanizedDocsis31ChannelStatus[]

  result.upstream = arrisDocsisStatus.upstream
    .filter(upstream => upstream.ChannelType === 'SC-QAM')
    .map(normalizeChannelStatus) as HumanizedDocsisChannelStatus[]

  result.upstreamOfdma = arrisDocsisStatus.upstream
    .filter(upstream => upstream.ChannelType === 'OFDMA')
    .map(normalizeChannelStatus) as HumanizedDocsis31ChannelStatus[]
  return result
}

export class Arris extends Modem {
  private csrfNonce = ''
  constructor(readonly modemIp: string, readonly logger: Log) {
    super(modemIp, logger)
  }

  async logout(): Promise<void> {
    try {
      this.logger.log('Logging out...')
      return  this.httpClient.post('/php/logout.php')
    } catch (error) {
      this.logger.error('Could not do a full session logout', error)
      throw error
    }
  }

  async login(password: string): Promise<void> {
    const cryptoVars = await this.getCurrentCryptoVars()
    const encPw = this.encryptPassword(password, cryptoVars)
    this.logger.debug('Encrypted password: ', encPw)
    const serverSetPassword = await this.createServerRecord(encPw)
    this.logger.debug('ServerSetPassword: ', serverSetPassword)

    const csrfNonce = this.loginPasswordCheck(
      serverSetPassword.encryptData,
      cryptoVars,
      deriveKey(password, cryptoVars.salt)
    )
    this.logger.debug('Csrf nonce: ', csrfNonce)

    await this.addCredentialToCookie()
    this.csrfNonce = csrfNonce
  }

  async  getCurrentCryptoVars(): Promise<CryptoVars> {
    try {
      const {data} = await this.httpClient.get('/', {
        headers: {Accept: 'text/html,application/xhtml+xml,application/xml'},
      })
      const cryptoVars = extractCryptoVars(data as string)
      this.logger.debug('Parsed crypto vars: ', cryptoVars)
      return cryptoVars
    } catch (error) {
      this.logger.error('Could not get the index page from the router', error)
      throw error
    }
  }

  encryptPassword(
    password: string,
    cryptoVars: CryptoVars
  ): SetPasswordRequest {
    const jsData =
    '{"Password": "' + password + '", "Nonce": "' + cryptoVars.sessionId + '"}'
    const key = deriveKey(password, cryptoVars.salt)
    const authData = 'loginPassword'
    const encryptData = encrypt(key, jsData, cryptoVars.iv, authData)

    return {
      EncryptData: encryptData,
      Name: Modem.USERNAME,
      AuthData: authData,
    }
  }

  loginPasswordCheck(
    encryptedData: string,
    cryptoVars: CryptoVars,
    key: string
  ): string {
    const csrfNonce = decrypt(key, encryptedData, cryptoVars.iv, 'nonce')
    return csrfNonce
  }

  async  createServerRecord(
    setPasswordRequest: SetPasswordRequest
  ): Promise<SetPasswordResponse> {
    try {
      const {data} = await this.httpClient.post<SetPasswordResponse>(
        '/php/ajaxSet_Password.php',
        setPasswordRequest
      )
      // TODO handle wrong password case
      // { p_status: 'Lockout', p_waitTime: 1 }
      if (data.p_status === 'Lockout') {
        throw new Error(`Remote user locked out for: ${data.p_waitTime}s`)
      }
      return data
    } catch (error) {
      this.logger.error('Could not pass password on remote router.', error)
      throw error
    }
  }

  async addCredentialToCookie(): Promise<void> {
    const credential = await this.fetchCredential()
    this.logger.debug('Credential: ', credential)
    // set obligatory static cookie
    this.cookieJar.setCookie(`credential= ${credential}`, `http://${this.modemIp}`)
  }

  async  fetchCredential(): Promise<string> {
    try {
      const {data} = await this.httpClient.get('/base_95x.js')
      return extractCredentialString(data as string)
    } catch (error) {
      this.logger.error('Could not fetch credential.', error)
      throw error
    }
  }

  async docsis(): Promise<DocsisStatus> {
    if (!this.csrfNonce) {
      throw new Error('A valid csrfNonce is required in order to query the modem.')
    }
    try {
      const {data} = await this.httpClient.get('/php/status_docsis_data.php', {
        headers: {
          csrfNonce: this.csrfNonce,
          Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
          Connection: 'keep-alive',
        },
      })
      return normalizeDocsisStatus(extractDocsisStatus(data as string))
    } catch (error) {
      this.logger.error('Could not fetch remote docsis status', error)
      throw error
    }
  }

  async restart(): Promise<unknown> {
    try {
      const {data} = await this.httpClient.post(
        'php/ajaxSet_status_restart.php',
        {
          RestartReset: 'Restart',
        },
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
            Connection: 'keep-alive',
          },
        }
      )
      this.logger.log('Router is restarting')
      return data
    } catch (error) {
      this.logger.error('Could not restart router.', error)
      throw error
    }
  }

  async firewall(status:boolean = true): Promise<void> {
    const action = status ? 'on' : 'off'
    try {
      const {data} = await this.httpClient.post(
        'api/v1/firewall',
        {
          FirewallLevel: action,
          FirewallLevelV6: action
        },
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
            Connection: 'keep-alive',
          },
        }
      )
      this.logger.log('Firewall is enabled')
      return data
    } catch (error) {
      this.logger.error('Could not enable firewall.', error)
      throw error
    }
  }

  _convertGetExposedHostSettings(settings: ArrisGetExposedHostSettings): ExposedHostSettings {
    return {
      serviceName: settings.ServiceName,
      mac: settings.MAC,
      protocol: settings.Protocol,
      startPort: settings.StartPort,
      endPort: settings.EndPort,
      enabled: settings.Status === "Enabled" ? true : false,
      index: Number.parseInt(settings.Index),
    } as ExposedHostSettings
  }

  async getHostExposure(): Promise<HostExposureSettings> {
    try {
      const {data} = await this.httpClient.get(
        'php/net_ipv6_host_exposure_data.php?{"hostExposure":{},"dhcpclient":{}}',
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?net_ipv6_host_exposure&mid=NetIPv6HostExposure`,
            Connection: 'keep-alive',
          },
        }
      )
      return {
        hosts: (data as ArrisGetHostExposureSettings)
          .hostExposure.map(this._convertGetExposedHostSettings)
      } as HostExposureSettings
    }
    catch (error) {
      this.logger.error("Could not get host exposure data:\n", error)
      throw error
    }
  }

  _convertSetExposedHostSettings(settings: ExposedHostSettings): ArrisSetExposedHostSettings {
    return {
      name: settings.serviceName,
      macAddress: settings.mac,
      protocol: settings.protocol,
      startPort: settings.startPort,
      endPort: settings.endPort,
      enable: settings.enabled ? "Enabled" : "Disabled",
      index: settings.index.toString(),
    }
  }

  async setHostExposure(settings: HostExposureSettings): Promise<void> {
    const convertedSettings =
      {hEditRule: settings.hosts.map(this._convertSetExposedHostSettings)} as ArrisSetHostExposureSettings
    try {
      await this.httpClient.post(
        'php/ajaxSet_net_ipv6_host_exposure_data.php',
        convertedSettings,
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?net_ipv6_host_exposure&mid=NetIPv6HostExposure`,
            Connection: 'keep-alive',
          }
        }
      )
    }
    catch (error) {
      console.error("Could not set host exposure data:\n", error)
      throw error
    }
  }
}

