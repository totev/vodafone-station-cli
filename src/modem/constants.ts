export const BAD_MODEM_POWER_LEVEL = -100
