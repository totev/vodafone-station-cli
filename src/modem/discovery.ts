import axios from "axios";
import { Log } from "../logger";
import { TechnicolorConfiguration } from "./technicolor-modem";
import { extractFirmwareVersion } from "./tools/html-parser";
const BRIDGED_MODEM_IP = "192.168.100.1";
const ROUTER_IP = "192.168.0.1";
axios.defaults.timeout = 10000;

export async function discoverModemIp(): Promise<string> {
  try {
    const maybeResult = await Promise.any([
      axios.get(`http://${BRIDGED_MODEM_IP}`),
      axios.get(`http://${ROUTER_IP}`),
    ]);
    if (maybeResult?.request?.host) {
      return maybeResult?.request?.host;
    }
    throw new Error();
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Could not find a router/modem under the known addresses.");
    throw error;
  }
}

export interface ModemInformation {
  deviceType: "Arris" | "Technicolor";
  firmwareVersion: string;
  ipAddress: string;
}

export class ModemDiscovery {
  constructor(private readonly modemIp: string, private readonly logger: Log) {}

  async tryTechnicolor(): Promise<ModemInformation> {
    const { data } = await axios.get<TechnicolorConfiguration>(
      `http://${this.modemIp}/api/v1/login_conf`
    );
    this.logger.debug(
      `Technicolor login configuration: ${JSON.stringify(data)}`
    );
    if (data.error === "ok" && data.data?.firmwareversion) {
      return {
        deviceType: "Technicolor",
        firmwareVersion: data.data.firmwareversion,
        ipAddress: this.modemIp,
      };
    }
    throw new Error("Could not determine modem type");
  }

  async tryArris(): Promise<ModemInformation> {
    const { data } = await axios.get(`http://${this.modemIp}/index.php`, {
      headers: {
        Accept: "text/html,application/xhtml+xml,application/xml",
      },
    });
    const firmwareVersion = extractFirmwareVersion(data as string);
    if (!firmwareVersion) {
      throw new Error("Unable to parse firmware version.");
    }
    return {
      deviceType: "Arris",
      firmwareVersion,
      ipAddress: this.modemIp,
    };
  }

  async discover(): Promise<ModemInformation> {
    try {
      const maybeModem = await Promise.any([
        this.tryArris(),
        this.tryTechnicolor(),
      ]);
      if (!maybeModem) {
        throw new Error("Modem discovery was unsuccessful");
      }
      return maybeModem;
    } catch (error) {
      this.logger.warn(
        "Could not find a router/modem under the known addresses"
      );
      throw error;
    }
  }
}
