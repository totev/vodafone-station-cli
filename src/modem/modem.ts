import axios, { AxiosInstance } from 'axios';
import { wrapper } from 'axios-cookiejar-support';
import { CookieJar } from 'tough-cookie';
import { Log } from '../logger';

export type DocsisChannelType = 'OFDM' | 'OFDMA' | 'SC-QAM'

export type Modulation = "16QAM" | "64QAM" | "256QAM" | "1024QAM" | "2048QAM" | "4096QAM"

export interface HumanizedDocsisChannelStatus {
  channelId: string;
  channelType: DocsisChannelType;
  snr: number; // dB
  frequency: number; // MHz
  modulation: Modulation;
  lockStatus: string;
  powerLevel: number; // dBmV
}

export interface DiagnosedDocsisChannelStatus extends HumanizedDocsisChannelStatus {
  diagnose: Diagnose
}
export interface DiagnosedDocsis31ChannelStatus extends HumanizedDocsis31ChannelStatus {
  diagnose: Diagnose
}

export interface Diagnose {
  deviation: boolean
  color: "red" | "green" | "yellow"
  description: string;
}

export interface HumanizedDocsis31ChannelStatus extends Omit<HumanizedDocsisChannelStatus, 'frequency'> {
  frequencyStart: number;// MHz
  frequencyEnd: number;// MHz
}

export interface DocsisStatus {
  downstream: HumanizedDocsisChannelStatus[];
  downstreamOfdm: HumanizedDocsis31ChannelStatus[];
  upstream: HumanizedDocsisChannelStatus[];
  upstreamOfdma: HumanizedDocsis31ChannelStatus[];
  time: string;
}

export interface DiagnosedDocsisStatus {
  downstream: DiagnosedDocsisChannelStatus[];
  downstreamOfdm: DiagnosedDocsis31ChannelStatus[];
  upstream: DiagnosedDocsisChannelStatus[];
  upstreamOfdma: DiagnosedDocsis31ChannelStatus[];
  time: string;
}

export interface GenericModem {
  logout(): Promise<void>;
  login(password: string): Promise<void>;
  docsis(): Promise<DocsisStatus>;
  restart(): Promise<unknown>;
}

export abstract class Modem implements GenericModem {
  protected readonly cookieJar: CookieJar
  protected readonly httpClient: AxiosInstance
  static USERNAME = 'admin'

  constructor(protected readonly modemIp: string, protected readonly logger: Log) {
    this.cookieJar = new CookieJar()
    this.httpClient = this.initAxios()
  }
  restart(): Promise<unknown> {
    throw new Error('Method not implemented.')
  }

  docsis(): Promise<DocsisStatus> {
    throw new Error('Method not implemented.')
  }

  login(_password: string): Promise<void> {
    throw new Error('Method not implemented.')
  }

  logout(): Promise<void> {
    throw new Error('Method not implemented.')
  }

  private initAxios(): AxiosInstance {
    return wrapper(axios.create({
      withCredentials: true,
      jar: this.cookieJar,
      baseURL: `http://${this.modemIp}`,
      headers: {
        'X-Requested-With': 'XMLHttpRequest',
      },
      timeout: 45000
    }))
  }
}

export function normalizeModulation(modulation: string): Modulation {
  let normalizedModulation = modulation;
  if (modulation.match("/")) {
    return normalizeModulation(modulation.split("/")[0]);
  }
  if (modulation.match("-")) {
    normalizedModulation = modulation.split("-").join("");
  }
  if (modulation.match(" ")) {
    normalizedModulation = modulation.split(" ").join("");
  }
  return normalizedModulation.toUpperCase() as Modulation;
}