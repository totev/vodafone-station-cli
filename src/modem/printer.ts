import type { DocsisStatus, HumanizedDocsis31ChannelStatus, HumanizedDocsisChannelStatus } from "./modem";

export class TablePrinter {

  head = [
    "ID",
    "Ch. Type",
    "Modulation",
    "Power",
    "Frequency",
    " Lock status ",
    "SNR "]

  constructor(private docsisStatus: DocsisStatus) { }

  get spacedHead(): string[] {
    return this.head.map(header => ` ${header} `)
  }

  get lineSeparator(): string {
    const dash = '-';
    const plus = '+'
    const dashes = this.spacedHead.map(word => dash.repeat(word.length))
    return ["", ...dashes, ""].join(plus)
  }

  lineText(...words: Array<string | number>): string {
    const paddedWords = words.map((word, index) => {
      const headerWord = this.head[index];
      return String(word).padEnd(headerWord.length, " ")
    })
    return ["", ...paddedWords, ""].join(" | ").trim()
  }

  tableHeader(): string {
    return `${this.lineSeparator}\n${this.lineText(...this.head)}\n${this.lineSeparator}`
  }

  tableRow(...words: Array<string | number>): string {
    return `${this.lineText(...words)}\n${this.lineSeparator}`
  }

  docsisStatusToRow(rowObjects: HumanizedDocsisChannelStatus[]): string {
    return rowObjects?.map(channelStatus => {
      return [channelStatus.channelId, channelStatus.channelType, channelStatus.modulation, channelStatus.powerLevel, channelStatus.frequency, channelStatus.lockStatus, channelStatus.snr]
    })
      .map(rowValues => this.tableRow(...rowValues))
      .join("\n") ?? ""
  }

  docsis31StatusToRow(rowObjects: HumanizedDocsis31ChannelStatus[]): string {
    return rowObjects?.map(channelStatus => {
      return [channelStatus.channelId, channelStatus.channelType, channelStatus.modulation, channelStatus.powerLevel, `${channelStatus.frequencyStart}-${channelStatus.frequencyEnd}`, channelStatus.lockStatus, channelStatus.snr]
    })
      .map(rowValues => this.tableRow(...rowValues))
      .join("\n") ?? ""
  }

  print(): string {
    const header = this.tableHeader();
    const downstream = this.docsisStatusToRow(this.docsisStatus.downstream)
    const downstreamOfdm = this.docsis31StatusToRow(this.docsisStatus.downstreamOfdm)
    const upstream = this.docsisStatusToRow(this.docsisStatus.upstream)
    const upstreamOfdma = this.docsis31StatusToRow(this.docsisStatus.upstreamOfdma)

    return `
Downstream\n${header}\n${downstream}\n
Downstream OFDM\n${header}\n${downstreamOfdm}\n
Upstream\n${header}\n${upstream}\n
Upstream OFDMA\n${header}\n${upstreamOfdma}
`
  }
}