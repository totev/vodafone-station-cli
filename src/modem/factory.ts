import {Arris} from './arris-modem'
import {ModemInformation} from '../discovery'
import {ConsoleLogger, Log} from '../logger'
import {Modem} from '../modem'
import {Technicolor} from './technicolor-modem'

export function modemFactory(modemInfo: ModemInformation, logger: Log = new ConsoleLogger()): Modem {
  switch (modemInfo.deviceType) {
  case 'Arris':
    return new Arris(modemInfo.ipAddress, logger)
  case 'Technicolor':
    return new Technicolor(modemInfo.ipAddress, logger)
  default:
    throw new Error(`Unsupported modem ${modemInfo.deviceType}`)
  }
}
