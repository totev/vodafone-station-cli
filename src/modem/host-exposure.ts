import {Log} from '../logger'
import {discoverModemLocation, ModemDiscovery} from './discovery'
import {modemFactory} from './factory'

export async function toggleHostExposureEntries(toggle: boolean, entries: string[], password: string, logger: Log): Promise<void> {
  const modemLocation = await discoverModemLocation()
  const discoveredModem = await new ModemDiscovery(modemLocation, logger).discover()
  const modem = modemFactory(discoveredModem, logger)
  try {
    await modem.login(password)
    const settings = await modem.getHostExposure()

    let names = entries
    if (names.length === 0) {
      names = settings.hosts.map(host => host.serviceName)
    }

    for (const name of names) {
      const index = settings.hosts.findIndex(host => host.serviceName === name)
      if (index === -1) {
        logger.warn(`Entry with the name '${name}' does not exist.`)
      } else {
        settings.hosts[index].enabled = toggle
      }
    }

    await modem.setHostExposure(settings)
  } catch (error) {
    logger.error('Could not change host exposure settings on modem.', error)
    throw error
  } finally {
    await modem.logout()
  }
}
