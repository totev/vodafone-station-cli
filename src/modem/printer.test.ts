import { DocsisStatus, HumanizedDocsisChannelStatus } from "./modem";
import { TablePrinter } from "./printer";
import fixtureDocsisStatus from './__fixtures__/docsisStatus_normalized.json';


describe('TablePrinter', () => {
  const printer = new TablePrinter(fixtureDocsisStatus as any)

  test('spacedHead spaces with one space before and after string', () => {
    expect(printer.spacedHead).toHaveLength(printer.head.length)
    expect(printer.spacedHead.map(head => head.length)).toEqual(printer.head.map(head => head.length + 2))
  });

  test('lineSeparator', () => {
    const expected = "+----+----------+------------+-------+-----------+---------------+------+"
    expect(printer.lineSeparator).toEqual(expected)
  });

  test('lineText single word', () => {
    expect(printer.lineText('Test')).toEqual('| Test |')
  });

  test('lineText multi words', () => {
    expect(printer.lineText('Test', "Test2")).toEqual('| Test | Test2    |')
  });

  test('tableHeader', () => {
    const expeced = `
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+`.trim()
    expect(printer.tableHeader()).toEqual(expeced)
  });

  test('tableRow', () => {
    const expeced = `
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+`.trim()
    expect(printer.tableRow(...printer.head)).toEqual(expeced)
  });

  test('print', () => {
    expect(printer.print()).toMatchSnapshot()
  });

});

test('print with uncompleted lock ', () => {
  const expected = `
Downstream
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+


Downstream OFDM
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+


Upstream
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+


Upstream OFDMA
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+
| 9  | OFDMA    | 16QAM      | NaN   | 29.8-NaN  | Not Completed | 0    |
+----+----------+------------+-------+-----------+---------------+------+
`
  const printer = new TablePrinter({
    upstreamOfdma: [{
      channelId: "9",
      channelType: "OFDMA",
      modulation: "16QAM",
      powerLevel: NaN,
      frequencyStart: 29.8,
      frequencyEnd: NaN,
      lockStatus: "Not Completed",
      snr: 0

    }]
  } as DocsisStatus)
  
  expect(printer.print()).toStrictEqual(expected)
});