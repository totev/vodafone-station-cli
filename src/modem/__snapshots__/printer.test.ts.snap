// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePrinter print 1`] = `
"
Downstream
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+
| 1  | SC-QAM   | 256QAM     | 55.1  | 114       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 2  | SC-QAM   | 256QAM     | 54.7  | 130       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 3  | SC-QAM   | 256QAM     | 54.8  | 138       | Locked        | 36   |
+----+----------+------------+-------+-----------+---------------+------+
| 4  | SC-QAM   | 256QAM     | 54.6  | 146       | Locked        | 36   |
+----+----------+------------+-------+-----------+---------------+------+
| 5  | SC-QAM   | 256QAM     | 57    | 602       | Locked        | 38   |
+----+----------+------------+-------+-----------+---------------+------+
| 6  | SC-QAM   | 256QAM     | 57.3  | 618       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 7  | SC-QAM   | 256QAM     | 57.7  | 626       | Locked        | 38   |
+----+----------+------------+-------+-----------+---------------+------+
| 8  | SC-QAM   | 256QAM     | 58.5  | 642       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 9  | SC-QAM   | 256QAM     | 58.3  | 650       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 10 | SC-QAM   | 256QAM     | 58.3  | 658       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 11 | SC-QAM   | 256QAM     | 58.1  | 666       | Locked        | 38   |
+----+----------+------------+-------+-----------+---------------+------+
| 12 | SC-QAM   | 256QAM     | 58.8  | 674       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 13 | SC-QAM   | 256QAM     | 58.8  | 682       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 14 | SC-QAM   | 256QAM     | 59.4  | 690       | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+
| 15 | SC-QAM   | 64QAM      | 53    | 698       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 16 | SC-QAM   | 64QAM      | 54.1  | 706       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 17 | SC-QAM   | 64QAM      | 54.2  | 714       | Locked        | 34   |
+----+----------+------------+-------+-----------+---------------+------+
| 18 | SC-QAM   | 64QAM      | 53.8  | 722       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 19 | SC-QAM   | 64QAM      | 53.9  | 730       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 20 | SC-QAM   | 64QAM      | 54.9  | 738       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 21 | SC-QAM   | 64QAM      | 55.3  | 746       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 22 | SC-QAM   | 64QAM      | 54.5  | 754       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 23 | SC-QAM   | 64QAM      | 54.5  | 762       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 24 | SC-QAM   | 64QAM      | 54.5  | 770       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 25 | SC-QAM   | 64QAM      | 55    | 778       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 26 | SC-QAM   | 64QAM      | 55    | 786       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 27 | SC-QAM   | 64QAM      | 54.9  | 794       | Locked        | 34   |
+----+----------+------------+-------+-----------+---------------+------+
| 28 | SC-QAM   | 64QAM      | 54.4  | 802       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 29 | SC-QAM   | 64QAM      | 54.1  | 810       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 30 | SC-QAM   | 64QAM      | 54.5  | 818       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 31 | SC-QAM   | 64QAM      | 54.5  | 826       | Locked        | 35   |
+----+----------+------------+-------+-----------+---------------+------+
| 32 | SC-QAM   | 64QAM      | 53.7  | 834       | Locked        | 34   |
+----+----------+------------+-------+-----------+---------------+------+

Downstream OFDM
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+
| 33 | OFDM     | 1024QAM    | 56.1  | 151-324   | Locked        | 39   |
+----+----------+------------+-------+-----------+---------------+------+

Upstream
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+
| 3  | SC-QAM   | 64QAM      | 110.3 | 37        | Locked        | 0    |
+----+----------+------------+-------+-----------+---------------+------+
| 4  | SC-QAM   | 32QAM      | 110.3 | 31        | Locked        | 0    |
+----+----------+------------+-------+-----------+---------------+------+
| 1  | SC-QAM   | 64QAM      | 110.3 | 51        | Locked        | 0    |
+----+----------+------------+-------+-----------+---------------+------+
| 2  | SC-QAM   | 64QAM      | 110.3 | 45        | Locked        | 0    |
+----+----------+------------+-------+-----------+---------------+------+

Upstream OFDMA
+----+----------+------------+-------+-----------+---------------+------+
| ID | Ch. Type | Modulation | Power | Frequency |  Lock status  | SNR  |
+----+----------+------------+-------+-----------+---------------+------+
| 9  | OFDMA    | 16_QAM     | 106.2 | 29.8-64.8 | SUCCESS       | 0    |
+----+----------+------------+-------+-----------+---------------+------+
"
`;
